core/src/host_imports/imports.rs

```diff
@@ -22,124 +22,188 @@ use super::lowering::LoHelper;
+    let memory = caller
+        .memory(memory_index)
+        .unwrap_or_else(|| panic!("Host import called directly, not from wasm"));
+
+    let inputs = lift_inputs::<WB>(
+        &mut caller,
+        memory.clone(),
+        record_types,
+        inputs,
+        argument_types,
+    );
+    let output = match inputs {
+        Ok(ivalues) => host_exported_func(&mut caller, ivalues),
+        Err(e) => {
+            log::error!("error occurred while lifting values in host import: {}", e);
+            error_handler
+                .as_ref()
+                .map_or_else(|| default_error_handler(&e), |h| h(&e))
+        }
     };
 
-    let raw_args = itypes_args_to_wtypes(&argument_types);
-    let raw_output = itypes_output_to_wtypes(&output_type_to_types(output_type));
-
-    let func = move |ctx: &mut Ctx, inputs: &[WValue]| -> Vec<WValue> {
-        let result = {
-            let memory_index = 0;
-            let memory = ctx.memory(memory_index);
-            let memory_view = WITMemoryView(memory.view::<u8>());
-            let li_helper = LiHelper::new(record_types.clone());
-            let lifter = ILifter::new(memory_view, &li_helper);
-
-            match wvalues_to_ivalues(&lifter, inputs, &argument_types) {
-                Ok(ivalues) => host_exported_func(ctx, ivalues),
-                Err(e) => {
-                    log::error!("error occurred while lifting values in host import: {}", e);
-                    error_handler
-                        .as_ref()
-                        .map_or_else(|| default_error_handler(&e), |h| h(&e))
-                }
-            }
-        };
-
-        init_wasm_func_once!(allocate_func, ctx, (i32, i32), i32, ALLOCATE_FUNC_NAME, 2);
-
-        let memory_index = 0;
-        let memory = ctx.memory(memory_index);
-        let memory_view = WITMemoryView(memory.view::<u8>());
-        let lo_helper = LoHelper::new(&allocate_func, ctx);
-        let t = ILowerer::new(memory_view, &lo_helper)
+    lower_outputs::<WB>(&mut caller, memory, output)
+}
+
+fn lift_inputs<WB: WasmBackend>(
+    caller: &mut <WB as WasmBackend>::Caller<'_>,
+    memory: <WB as WasmBackend>::Memory,
+    record_types: Arc<MRecordTypes>,
+    inputs: &[WValue],
+    argument_types: &[IType],
+) -> HostImportResult<Vec<IValue>> {
+    let memory_view = memory.view();
+    let li_helper = LiHelper::new(record_types);
+    let lifter = ILifter::new(memory_view, &li_helper);
+    wvalues_to_ivalues(
+        &mut caller.as_context_mut(),
+        &lifter,
+        inputs,
+        argument_types,
+    )
+}
+
+fn lower_outputs<WB: WasmBackend>(
+    caller: &mut <WB as WasmBackend>::Caller<'_>,
+    memory: <WB as WasmBackend>::Memory,
+    output: Option<IValue>,
+) -> Vec<WValue> {
+    init_wasm_func!(
+        allocate_func,
+        caller,
+        (i32, i32),
+        i32,
+        ALLOCATE_FUNC_NAME,
+        2
+    );
+
+    let is_record = matches!(&output, Some(IValue::Record(_)));
```

Comment:
let's define it closer to the place of usage

Reply:
The value is moved at statment starting at line 134. I can move it closer if i add an unnecessary clone. Is it worth it?